joey/layers_v2.py

```diff
@@ -391,3 +393,357 @@ def __init__(self, kernel_size, input_size,
         super().__init__(kernel_size, input_size, dimensions,
                          stride, padding, activation, generate_code,
                          strict_stride_check)
+
+
+class Pooling(Layer):
+    """
+    A Layer abstract subclass corresponding to a generic pooling layer.
+    When you create a subclass of Pooling, you have to implement
+    the following methods: equations(), backprop_equations().
+
+    Parameters
+    ----------
+    kernel_size : (int, int)
+        The shape of a kernel (represented internally by a NumPy array)
+        expressed as (rows, columns).
+    input_size : (int, int, int, int)
+        The shape of input data expressed as
+        (batch size, channels, rows, columns).
+    name_allocator_func : zero-argument function, optional
+        See Layer.__doc__.
+    dim_allocator_func : one-argument function, optional
+        See Layer.__doc__.
+    stride : (int, int), optional
+        Stride of the layer expressed as (rows, columns). The default
+        value is (1, 1).
+    padding : (int, int), optional
+        Padding of the layer expressed as (rows, columns). The default
+        value is (0, 0).
+
+        Be careful! The current version of Joey supports non-zero padding
+        ONLY for standalone layers. When you create a neural network, all
+        of its layers must have (0, 0) padding.
+    activation : Activation, optional
+        See Layer.__doc__. The actual default value is Dummy.
+    generate_code : bool, optional
+        See Layer.__doc__.
+    strict_stride_check : bool, optional
+        A boolean indicating whether a strict stride check should be
+        performed when instantiating this object. The default value is
+        True.
+
+        If the check is disabled and the stride turns out to be
+        incompatible with the provided kernel, input and padding sizes,
+        some parts of input data will not be processed. This behaviour
+        is intentional, its aim is avoiding any out-of-bounds accesses.
+    """
+
+    def __init__(self, kernel_size, input_size,  dimensions,
+                 name_allocator_func=alloc, dim_allocator_func=dim_alloc,
+                 stride=(1, 1), padding=(0, 0), activation=None,
+                 generate_code=False, strict_stride_check=True):
+        # Kernel size is expressed as (rows, columns).
+        # Input size is expressed as (batch size, channels, rows, columns).
+        self._dims = dimensions
+        self._dim_dict = {3: 'depth', 2: 'height', 1: 'width'}
+
+        if (type(padding) is int):
+            padding = tuple([padding] * self._dims)
```

Comment:
We should properly handle the else cases of both this and the following if. e.g. add an assert

Reply:
I was trying to replicate pyTorch interface where a user can either specify the padding/stride as a tuple or a single number, the assertion is handled in the _error_check() method if its not a single number

Reply:
Important to do this.